### ISSUE TITLE
Deno mixes JS and TS if shebang is used but no file extension

### ISSUE THREAD

**Issue by Brixy** [denoland/deno#17195]
Hi, I use Deno for OS scripting. Scripting files use shebangs but no file extension, e.g.

```
// my_script
#!/bin/env -S deno run --allow-env --allow-read
```

When I set the `shebangs` setting for JavaScript (but not for TypeScript) the language is correctly recognized as JavaScript.

```
# languages.toml
[[language]]
name = "javascript"
shebangs = ["deno", "node"]
auto-format = true
config = { enable = true, lint = true, unstable = true }
language-server = { command = "deno", args = ["lsp"], language-id = "javascript" }

[[language]]
name = "typescript"
# shebangs = ["deno", "node"]
auto-format = true
config = { enable = true, lint = true, unstable = true }
language-server = { command = "deno", args = ["lsp"], language-id = "typescript" }
```

But error messages are TypeScript specific. So there is is strange mix of JavaScript and TypeScript functionalities.

As soon as the file is renamed to `my_script.js` (or `my_script.ts`) everything works as expected.

Also, if I wanted to use JS and TS with shebangs only there is no way the differentiate between the two (if I get it correctly). 

Maybe there should be a way to set the language in the shebang line, e.g. with

```
#!/bin/env -S deno run --allow-env --allow-read --language=javascript
```

Thanks a lot!

**Comment by sigmaSd** [comment]
seems like deno lsp defaults to typescript and deno cli defaults to javascript

I remember there was an issue to add --ext=js/ts

Personally I prefer using `deno install -n namewithoutext file.js`

**Comment by sigmaSd** [comment]
The current workaround is to use typescript, even scripts needs typing xD

**Comment by Brixy** [comment]
Thanks a lot for your tips!

> The current workaround is to use typescript, even scripts needs typing xD

Yes, using TypeScript is no problem (or preferable) for scriping for OSs.

I just wanted to report this in case it this phenomenon is a bug.

**Comment by ayame113** [comment]
FYI, this issue seems to have been discussed in #5088 for a long time. Hopefully #17172 will resolve it I think.

**Comment by sigmaSd** [comment]
That issue is a bit different, and I think ppl have moved on because deno run defaults to no --no-check now

The issue here is the lsp, it defaults to typescript even if you send a language I'd=js

I didn't look how the lsp spec says to handle this case, but I'm guessing it's incorrect

**Comment by ayame113** [comment]
Oh, I didn't realize you were talking about lsp here. Sorry for the noise!😓

**Comment by sigmaSd** [comment]
yeah its not clear from the issue, its just what I figured out after debugging

**Comment by sigmaSd** [comment]
This is what the spec says https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem

![image](https://user-images.githubusercontent.com/22427111/210046506-ac628ec3-6933-406f-8799-11931338c35f.png)

So the spec indeed says that the server should not re interpret the language extension, and it should use the language id instead

**Comment by sigmaSd** [comment]
I followed the code a bit in github , and it seems that deno lsp detects correctly  the language id, but either way it will generate the typescipt diagnostics irregardless of it https://github.com/denoland/deno/blob/218d7ab7780d5e44d3badd1102a65a1be2afd00a/cli/lsp/diagnostics.rs#L236

**Comment by Avimitin** [comment]
This issue can cause another problem. deno-run default use JavaScript engine to interpret the script file without `.ts` extension, which leads to a problem that it cannot correctly parse script with TypeScript specific content. For example:

```txt
$ echo "interface A {}" > test
$ echo "interface A {}" > test.ts
$ deno run test
error: The module's source code could not be parsed: Expression expected at file:///home/sh1marin/w/avimitin/infra/scripts/archrv/test:1:1

  interface A {}
  ~~~~~~~~~
$ deno run test.ts
# OK
```

**Comment by Avimitin** [comment]
Gcc compiler have a `-x` option to determine input language, maybe deno can implement something like that:

```text
GCC(1)
-x language
           Specify explicitly the language for the following input files
           (rather than letting the compiler choose a default based on the
           file name suffix).  This option applies to all following input
           files until the next -x option.
```

**Comment by beetcb** [comment]
https://docs.deno.com/runtime/manual/advanced/typescript/overview#determining-the-type-of-file
https://docs.deno.com/runtime/manual/advanced/typescript/overview#supported-media-types

> The module type can be overridden using the `--ext` argument. This is useful if the module does not have a file extension, e.g. because it is embedded in another file.

and 

> For remote modules, the media type (mime-type) is used to determine the type of the module, where the path of the module is used to help influence the file type, when it is ambiguous what type of file it is.

**Comment by simonjur** [comment]
With shebang like
```
#!/usr/bin/env -S deno run --ext ts --allow-read --allow-sys --allow-env --allow-run --allow-write
```
script works again

**Comment by littledivy** [comment]
The CLI side of this is resolved. The `--ext` flag was added (#17172), and extensionless scripts containing TypeScript syntax now run correctly — `deno run scriptfile` with `interface A {}` works, and `#!/usr/bin/env -S deno run --ext ts ...` in a shebang works as expected. The remaining concern was the LSP defaulting to TypeScript diagnostics even when the client sends `language-id: javascript`; if that's still a problem, a focused issue specifically about LSP language-id handling would track it more clearly. Closing the CLI portion as resolved.
